enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,319 @@
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Glossary
+
+### Lifecycle Hook
+A specific point in a machine's reconciliation lifecycle where execution of
+normal machine-controller behaviour is paused or modified.
+
+### Deletion Phase
+Describes when a machine has been marked for deletion but is still present
+in the API.  Various actions happen during this phase, such as draining a node,
+deleting an instance from a cloud provider, and deleting the node object.
+
+### Hook Implementing Controller (HIC)
+The Hook Implementing Controller describes a controller, other than the
+machine controller, that adds, removes, and/or responds to a particular
+lifecycle hook. Each lifecycle hook should have a single HIC, but an HIC
+can optionally manage one or more hooks.
+
+## Summary
+
+Defines a set of annotations that can be applied to a Machine which affect the
+linear progress of a Machine’s lifecycle after a Machine has been marked for
+deletion. These annotations are optional and may be applied during Machine
+creation, sometime after Machine creation by a user, or sometime after Machine
+creation by another controller or application.
+
+## Motivation
+
+Allow custom and 3rd party components to easily interact with a Machine or
+related resources while that Machine's reconciliation is temporarily paused.
+This pause in reconciliation will allow these custom components to take action
+after a Machine has been marked for deletion, but prior to the Machine being
+drained and/or associated instance terminated.
+
+In particular, this will allow the Etcd Operator to prevent a Control Plane
+Machine from being drained/removed until the Etcd Operator has had a chance
+to synchronise the Etcd data to the replacement Machine.
+
+This could also be used by other components such as Storage to allow them to
+ensure safe detachment of volumes before the Machine is deleted.
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+### Non-Goals
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases.
+- Create a mechanism to signal what lifecycle point a machine is at currently.
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine controller.
+- Require anyone to use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
```

Comment:
I don't personally have concerns about using annotations, we discussed a number of options when implementing this upstream and it seemed appropriate and safe to use annotations. There are plenty of projects and APIs with Kubernetes that rely on annotations in some way or another, for example a lot of the service controller implementations across clouds add extra capability via annotations, so I think there's prior art here.

> What is the failure case if the user or X removed an expected annotation?

Not sure I follow this question? Are you asking what happens when a user removes one of the hook annotations, or if they mistakenly remove more annotations other than the one they intended?

> Today who has permissions to update machines is this only cluster-admin?

By default yes, cluster admins have access to the modify machines, plus anything/anyone that the admins delegate this responsibility too

Reply:
I think to a degree this is intended behaviour. For example in a scenario where there is no way to add new capacity (restricted environments/at quota for cloud), then having the ability for a user to manually intervene and allow the less safe approach will be required, even in the etcd case. I think we need to account for these on the hook implementing controller side - the hooks, like finalizers, should be considered a convenience but not a guarantee

Reply:
Shouldn't the controller that needs the hook notice that there is a Machine without the annotation that is not being deleted and restore the missing settings?

Reply:
In the case that a user interferes and the Machine isn't being deleted, then yes, I would expect the hook implementing controller to re-add the hook and re-instate the protection.

What I think we need to avoid is if a user manually removes the hook while the Machine is being deleted, re-adding the hook at this point. 

To allow users to have an escape hatch when there are scenarios that require it, we need to make sure that the user can override the mechanism to allow the Machine API to continue with the removal even when the hook implementing controller deems it unsafe.

If the user removes it, and the hook implementing controller immediately re-adds it, then that could cause issues (and a game of hook ping pong) that then prevent the user being able to remove the Machine.

Perhaps having the escape hatch within the hooks themselves isn't the appropriate place, I guess the alternative is to have something which the implementing operator would observe and then remove the hook?

In the control plane case, we could have something on the CRD that we plan to implement that signals to the etcd operator the alternative cases that we need to cover (and therefore wouldn't need an escape hatch). I'm thinking something like rolloutStrategies inspired by those in statefulsets, defaulting to bursting, but also allowing "recreate"

Reply:
If controllers do not add their hooks to resources where the delete timestamp is not nil (just like with finalizers), everything should work fine. The hook will be restored if it is removed during normal operation and it won't be restored if it is removed during deletion, so the user will have a way to bypass the hook.

Reply:
I've already captured this in the partner enhancement that talks about how Etcd will leverage this feature, I'll make sure it's included here in the risks/mitigations section